prog/dftb+/lib_common/timer.F90

```diff
@@ -0,0 +1,130 @@
+module timer
+  use globalenv, only : stdOut
+  implicit none
+  private
+
+  public :: TTimer
+
+  !> Simple timer object
+  type :: TTimer
+    private
+    real :: startTime = 0.0
+    real :: endTime = -1.0
+    integer :: startCount = 0
+    integer :: endCount = -1
+    integer :: countRate = 1
+  contains
+    procedure :: start
+    procedure :: stop => stopTimer
+    procedure :: getCpuTime
+    procedure :: getWallClockTime
+    procedure :: writeTimes
+  end type TTimer
+
+contains
+
+  !> Starts the timer.
+  subroutine start(this)
+
+    !> Instance.
+    class(TTimer), intent(inout)   :: this
+
+    call cpu_time(this%startTime)
+    call system_clock(count=this%startCount, count_rate=this%countRate)
+
+  end subroutine start
+
+
+  !> Stops the timer.
+  subroutine stopTimer(this)
+
+    !> Instance.
+    class(TTimer), intent(inout) :: this
+
+    call cpu_time(this%endTime)
+    call system_clock(count=this%endCount)
+
+  end subroutine stopTimer
+
+
+  !> Returns the measured CPU time
+  function getCpuTime(this) result(cpuTime)
+
+    !> Instance.
+    class(TTimer), intent(in) :: this
+
+    !> Cpu time evolved between the last start() and stop() calls.
+    real :: cpuTime
+
+    real :: endTime
+
+    if (this%endTime < 0.0) then
+      call cpu_time(endTime)
+    else
+      endTime = this%endTime
+    end if
+    cpuTime = endTime - this%startTime
+
+  end function getCpuTime
+
+
+  !> Returns the measured wall clock time.
+  function getWallClockTime(this) result(wallClockTime)
+
+    !> Instance.
+    class(TTimer), intent(in) :: this
+
+    !> Wall clock time evolved between the last start() and stop() calls.
+    real :: wallClockTime
+
+    integer :: endCount
+
+    if (this%endCount < 0) then
+      call system_clock(count=endCount)
+    else
+      endCount = this%endCount
+    end if
+    if (this%countRate == 0) then
+      wallClockTime = 0.0
+    else
+      wallClockTime = real(endCount - this%startCount) / real(this%countRate)
+    end if
+
+  end function getWallClockTime
+
+
+  !> Writes the current measured times.
+  subroutine writeTimes(this, msg, node, fp)
```

Comment:
Any reason not to set msg at initialisation?

Reply:
If you wanted to print timings for every scc-cycle using the same timer and add the scc-cycle number in the message, it could be handy to specify the msg only when you print. Also timerArray initialises a lot of timers, but never uses their `writeTimes()` functions, so passing a message at init-time would be meaningless.